File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.cpp
class Solution {
public:
    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(), position.end());
        int n=position.size(), s=1, e=(position[n-1]-position[0])/(m-1);
        int ans;
        while(s<=e){
            int p=0;
            int mid=(s+e)/2;
            int b=1;
            for(int i=1; i<n ; i++){
                if(b<m){
                    if(position[i]-position[p]>=mid){
                        p=i;
                        b++;
                    }
                }
                else{
                    break;
                }
            }
            if(b==m){
                ans=mid;
                s=mid+1;
            }
            else{
                e=mid-1;
            }
        }
        return ans;
    }
};

File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        
        def fn(d):
            """Return True if d is a feasible distance."""
            ans, prev = 0, -inf # where previous ball is put
            for x in position:
                if x - prev >= d: 
                    ans += 1
                    if ans == m: return True
                    prev = x
            return False 
        
		# "last True" binary search (in contrast to "first True" binary search)
        lo, hi = 1, position[-1] - position[0]
        while lo < hi: 
            mid = lo + hi + 1 >> 1
            if fn(mid): lo = mid
            else: hi = mid - 1
        return lo